### PR TITLE
Make sliced models HuggingFace compatible

### PR DESCRIPTION
**Pull request by LianaMikael**
This PR adds the implementations for sliced Phi and Llama models to make it easy to save and load sliced models. 
The models can be initialized with a given scheduler (or no scheduler for zero sparsity) and support `save_pretrained` and `from_pretrained` methods like standard HF models.

### CODE REVIEW

File: experiments/run_slicegpt.py [hunk 1]
```diff
@@ -8,9 +8,13 @@
 import shutil
 
 import torch
-import wandb
+from transformers.models.llama.modeling_llama import LlamaConfig
+from transformers.models.phi.modeling_phi import PhiConfig
 
+import wandb
 from slicegpt import data_utils, gpu_utils, hf_utils, layernorm_fusion, rotate, utils
+from slicegpt.adapters.hf_compatible_phi import SlicedPhiForCausalLM
```
Comment (pashminacameron) [inline experiments/run_slicegpt.py:16]: `SlicedPhiForCausalLM` is in `sliced_phi.py`. Looks like this file was renamed and the import is using the old name.

File: experiments/run_slicegpt.py [hunk 2]
```diff
@@ -8,9 +8,13 @@
 import shutil
 
 import torch
-import wandb
+from transformers.models.llama.modeling_llama import LlamaConfig
```
Comment (pashminacameron) [inline experiments/run_slicegpt.py:11]: I think we might be missing one abstraction. We shouldn't need any model specific imports here. `hf_utils.get_model_and_tokenizer` abstracts the model type away; we want the same for saving HF-compatible sliced models. This `save` abstraction will also avoid the `Sliced<Model>ForCausalLM` imports here.

File: src/slicegpt/adapters/sliced_phi.py [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+import torch.nn as nn
+from transformers.models.phi.modeling_phi import PhiConfig, PhiForCausalLM, PhiModel
+
+from slicegpt.adapters.phi2_adapter import CompressedPhiDecoderLayer, Phi2ModelAdapter
+from slicegpt.modules import RMSN
+from slicegpt.rotate import slice_rotated_model
+from slicegpt.slicing_scheduler import SlicingScheduler
+
+
+class SlicedPhi(PhiModel):
```
Comment (pashminacameron) [inline src/slicegpt/adapters/sliced_phi.py:11]: I wonder if we should change the `model_type` to `sliced_phi` along the lines of `quarot_phi` etc, but that requires us to extend the config class. Let's leave this for a future PR.

File: tests/test_slicing.py [hunk 4]
```diff
@@ -21,3 +26,75 @@ def test_layernorm_fusion_replaces_modules() -> None:
...
+def test_HF_model():
+    """Check that the HF model weights are equivalent to the sliced model weights"""
+    model_name = "microsoft/phi-2"
+    model_adapter, tokenizer = hf_utils.get_model_and_tokenizer(model_name)
+    sparsity = 0.1
+
+    layernorm_fusion.replace_layers(model_adapter)
+    layernorm_fusion.fuse_modules(model_adapter)
+
+    config = PhiConfig.from_pretrained(
```
Comment (pashminacameron) [inline tests/test_slicing.py:40]: If we need a full Phi2 for the test, then we need the pytest markers like the ones in `test_experiments.py`.

File: src/slicegpt/adapters/sliced_phi.py (hunk 3 above)
Comment (LianaMikael) [reply src/slicegpt/adapters/sliced_phi.py:23]: Added the configs and model type

File: src/slicegpt/adapters/sliced_phi.py [hunk 5]
```diff
@@ -0,0 +1,85 @@
...
+from slicegpt.modules import RMSN
+from slicegpt.rotate import slice_rotated_model
+from slicegpt.slicing_scheduler import SlicingScheduler
+
+
+class SlicedPhi2Config(PhiConfig):
+    model_type = "sliced_phi2"
+    is_composition = True
+
+    def __init__(self, sparsity=0.1, new_hidden_size=1024, **kwargs):
```
Comment (pashminacameron) [inline src/slicegpt/adapters/sliced_phi.py:15]: Shall we leave the arguments `sparsity` and `new_hidden_size` uninitialised? Otherwise it could create confusion as to where the value came from when not explicitly set. And maybe we want these to be mandatory arguments?

File: src/slicegpt/hf_utils.py [hunk 6]
```diff
@@ -173,3 +180,51 @@ def load_sliced_model(
...
+):
+    if model_name == "microsoft/phi-2":
+        config = PhiConfig.from_pretrained(
+            model_name,
+            torch_dtype=torch.float16,
+        )
+
+        config.save_pretrained("phi_config")
+        config_to_save = SlicedPhi2Config.from_pretrained(
+            config_path="phi_config", sparsity=sparsity, new_hidden_size=new_hidden_size
```
Comment (pashminacameron) [inline src/slicegpt/hf_utils.py:203]: I am curious how `config_path` works here? Is this a real path? Is it not a name of config recognized by HF or a local path? Or is it simply reading back the file saved in the line above? In which case it could, in principle, be saved anywhere, rather than the current directory, right? (Not asking for a change, rather seeking an explanation)

**Review by pashminacameron — APPROVED** [review]
Since there is a single intermediate size, we can make this explicit by using the type ConstSlicingScheduler in the interface instead of SlicingScheduler. Otherwise, we can add support for all slicing schedulers and keep the base type SlicingScheduler. I don't mind how this is done - in this PR or a follow up PR. We should update the PR title to reflect the decision and contents of the PR.

**Review by pashminacameron — COMMENTED** [review]
Hmm :/ Tests are not working due to import issues.

File: src/slicegpt/hf_utils.py (hunk 6 above)
Comment (LianaMikael) [reply src/slicegpt/hf_utils.py:202]: Here it's simply reading the file saved above, so we can save it anywhere. Although I think this way of doing this is not ideal. Another option is to copy over each parameter from that config to the new one but it's also not very convenient.  Ideally we want to have all the parameters initialized correctly by default in the config and only pass `sparsity `and `new_hidden_size`. I will think about how to do this better in a subsequent PR.

File: tests/test_model_adapter.py [hunk 7]
```diff
@@ -102,6 +102,7 @@ def create_adapter(self) -> OPTModelAdapter:
         config = OPTConfig(
             vocab_size=32,
             hidden_size=8,
+            intermediate_size=32,
```
Comment (pashminacameron) [inline tests/test_model_adapter.py:105]: Do we need this? OPTConfig does not have a property called `intermediate_size`.

File: src/slicegpt/adapters/opt_adapter.py [hunk 8]
```diff
@@ -177,6 +177,10 @@ def seqlen(self) -> int:
     def hidden_size(self) -> int:
         return self.config.hidden_size
 
+    @property
+    def intermediate_size(self) -> int:
+        return self.config.ffn_dim
```
Comment (pashminacameron) [inline src/slicegpt/adapters/opt_adapter.py:182]: I think this is the right property to use for OPT but still can't make the tests for vanilla OPT slicing work. @nailimixaM @LianaMikael

Comment (LianaMikael) [reply src/slicegpt/adapters/opt_adapter.py:182]: I initially added intermediate_size just so we can save it with the slicing config but now that we have a proper config with copied parameters from the original config, I think this isn't needed anymore. I can just remove it

**Comment by pashminacameron** [comment]
I would like to hold off merging this into main for a bit. I will work on this more (after the smaller fixes) and we can re-review.

**Comment by canamika27** [comment]
> I would like to hold off merging this into main for a bit. I will work on this more (after the smaller fixes) and we can re-review.

Any update by when the changes will be merged

File: src/slicegpt/hf_utils.py [hunk 9]
```diff
@@ -102,39 +107,65 @@ def get_model_and_tokenizer(
...
+        scheduler = ConstSlicingScheduler(new_embedding_dimension)
+
+        layers = model_adapter.get_layers()
+        scheduler.setup(
+            hidden_size=model_adapter.hidden_size,
+            layers_num=len(layers),
+            parallel_blocks=True,
+        )
+
+        sliced_model = SlicedPhiForCausalLM.from_pretrained(
```
Comment (pashminacameron) [inline src/slicegpt/hf_utils.py:154]: The `sliced_model` can be of type `SlicedPhiForCausalLM` or `SlicedLlamaForCausalLM` at this point. Debugging..

File: src/slicegpt/hf_utils.py [hunk 10]
```diff
@@ -102,39 +107,65 @@ def get_model_and_tokenizer(
     model.eval()  # This switches off dropout.
     model_adapter.use_cache = False
 
-    tokenizer = AutoTokenizer.from_pretrained(model_path, use_fast=True, token=token, local_files_only=local_model)
+    tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True, token=token, local_files_only=local_model)
```
Comment (msdmkats) [inline src/slicegpt/hf_utils.py:110]: This change breaks loading of local models.

Comment (msdmkats) [reply src/slicegpt/hf_utils.py:110]: Probably, should remain `model_path`.

**Comment by madhusrivatsav** [comment]
Hi

Please let me know when these changes will be merged ?
Has anyone verified if the models are are HF compatible now ?

**Comment by NamburiSrinath** [comment]
Hi @pashminacameron and @LianaMikael,

Is there a timeline to merge these changes to main?
